### PR TITLE
Authentification with EPFL people API

### PR DESCRIPTION
**Pull request by Tikipiou**

### CODE REVIEW

File: app/src/main/java/com/swent/assos/MainActivity.kt [hunk 1]
```diff
@@ -15,6 +17,16 @@ import dagger.hilt.android.AndroidEntryPoint
...
   override fun onCreate(savedInstanceState: Bundle?) {
     super.onCreate(savedInstanceState)
+    // if we are in debug mode, we want to use the firestore emulator
+    val config = Config()
+    if (config.get_demo()) {
+      // Configure Firestore to use the Firestore emulator
+
+      // val firestoreSettings = FirebaseFirestoreSettings.Builder()
+
+      // FirebaseFirestore.getInstance().firestoreSettings = firestoreSettings
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/MainActivity.kt:27]: Delete the commented code

File: app/src/main/java/com/swent/assos/model/data/User.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.swent.assos.model.data
+
+data class User(
+    val uid: String,
+    val firsname: String,
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/model/data/User.kt:5]: Typo: `firsname` should be replaced by `firstName`

File: app/src/main/java/com/swent/assos/model/data/User.kt
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/model/data/User.kt:1]: Consider using CamelCase convention: replace `lastname` by `lastName`

File: app/src/main/java/com/swent/assos/model/service/impl/AuthServiceImpl.kt [hunk 3]
```diff
@@ -4,13 +4,26 @@ import com.google.android.gms.tasks.Task
...
+  // TODO: if you are using the firestore emulator, make sure to change the host and port to match
+  // your firestore emulator
+  val config = Config()
+
+  // if config.get_demo() is true, we are in debug mode and we want to use the firestore emulator
+  val ok =
+      if (config.get_demo()) {
+        auth.useEmulator("10.0.2.2", 9099)
+      } else {}
+  // val ok = auth.useEmulator("10.0.2.2",9099)
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/model/service/impl/AuthServiceImpl.kt:25]: Delete commented code

File: app/src/main/java/com/swent/assos/model/service/impl/AuthServiceImpl.kt [hunk 4]
```diff
@@ -4,13 +4,26 @@ import com.google.android.gms.tasks.Task
...
+import com.swent.assos.config.Config
 import com.swent.assos.model.service.AuthService
 import javax.inject.Inject
 import kotlinx.coroutines.channels.awaitClose
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.callbackFlow
 
 class AuthServiceImpl @Inject constructor(private val auth: FirebaseAuth) : AuthService {
+
+  // TODO: if you are using the firestore emulator, make sure to change the host and port to match
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/model/service/impl/AuthServiceImpl.kt:16]: The keyword TODO is maybe not the best in this context, it is more an advice for future implementations. You could replace it by `NB`

File: app/src/main/java/com/swent/assos/model/view/LoginViewModel.kt [hunk 5]
```diff
@@ -13,6 +18,44 @@ constructor(private val storageService: DbService, private val accountService: A
...
+  fun hashEmail(email: String): Int {
+    val bytes = email.toByteArray()
+    val md = MessageDigest.getInstance("SHA-256")
+    val digest = md.digest(bytes)
+    // get the integer value
+    return digest.fold(0) { acc, byte -> acc * 256 + byte.toInt() }
+  }
+
+  // fun goToSignUp() = navController.navigate("SignUp")
+  // fun goToSignIn() = navController.navigate("Login")
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/model/view/LoginViewModel.kt:57]: Remove commented code

File: app/src/main/java/com/swent/assos/ui/login/LoginScreen.kt [hunk 6]
```diff
@@ -23,11 +23,12 @@ import androidx.compose.ui.text.input.PasswordVisualTransformation
 import androidx.compose.ui.text.style.TextDecoration
 import androidx.compose.ui.unit.dp
 import androidx.hilt.navigation.compose.hiltViewModel
+import androidx.navigation.NavHostController
 import com.swent.assos.R
 import com.swent.assos.model.view.LoginViewModel
 
 @Composable
-fun LoginScreen() {
+fun LoginScreen(navController: NavHostController) {
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/LoginScreen.kt:31]: replace `navController` by an argument with type `NavigationActions`. It is better for encapsulation (we limit the access to the NavHostController)

File: app/src/main/java/com/swent/assos/ui/login/LoginScreen.kt [hunk 7]
```diff
@@ -55,13 +56,20 @@ fun LoginScreen() {
...
+                navController.navigate("AssociationPage")
+              }
+            },
         ) {
           Text("Login")
         }
 
         Text(
-            modifier = Modifier.clickable {},
+            // if clikecd, go to sign up page using hilt navigation
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/LoginScreen.kt:71]: Typo: `clikecd` should become `clicked`

File: app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt [hunk 8]
```diff
@@ -9,42 +10,89 @@ import androidx.compose.runtime.getValue
...
+import com.google.firebase.functions.FirebaseFunctions
+import com.google.firebase.functions.functions
+import com.swent.assos.model.navigation.Destinations
 import com.swent.assos.model.view.LoginViewModel
+import org.w3c.dom.Text
 
-@Preview
 @Composable
-fun SignUpScreen() {
+fun SignUpScreen(navController: NavController) {
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt:27]: Same remark: consider using NavigationsActions

File: app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt [hunk 9]
```diff
@@ -9,42 +10,89 @@ import androidx.compose.runtime.getValue
...
+        onValueChange = { email = it },
+        label = { Text("Email") },
+        modifier = Modifier.semantics { testTag = "email" })
     OutlinedTextField(
         value = password,
         onValueChange = { password = it },
         label = { Text("Password") },
-        visualTransformation = PasswordVisualTransformation())
+        visualTransformation = PasswordVisualTransformation(),
+        modifier = Modifier.semantics { testTag = "password" })
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt:45]: We use a different way to write testTags in the rest of the project, it is maybe a good idea to write it in the same way in the entire project. `Modifier.semantics { testTag = "password" }` would become `Modifier.testTag("password")`

File: app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt [hunk 10]
```diff
@@ -9,42 +10,89 @@ import androidx.compose.runtime.getValue
...
     OutlinedTextField(
         value = confirmPassword,
         onValueChange = { confirmPassword = it },
         label = { Text("Confirm Password") },
-        visualTransformation = PasswordVisualTransformation())
+        visualTransformation = PasswordVisualTransformation(),
+        modifier = Modifier.semantics { testTag = "confirmPassword" })
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt:51]: Same remark:  `Modifier.semantics { testTag = "confirmPassword" }` would become `Modifier.testTag("confirmPassword")`

File: app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt [hunk 11]
```diff
@@ -9,42 +10,89 @@ import androidx.compose.runtime.getValue
...
     if (password != confirmPassword) {
       Text("Passwords do not match", color = Color.Red)
     }
     Button(
         onClick = {
           if (password == confirmPassword && password.isNotEmpty()) {
+            println(loginViewModel.firestoreInstance)
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt:58]: Delete debugging log

File: app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt [hunk 12]
```diff
@@ -9,42 +10,89 @@ import androidx.compose.runtime.getValue
...
             loginViewModel.signUp(email, password)
+
+            // call the firebasefunction -> oncallFind.py
+            val data = hashMapOf("email" to email)
+
+            val functions = FirebaseFunctions.getInstance("europe-west6")
+            // Firebase.functions.useEmulator("10.0.2.2", 5001)
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt:65]: Delete commented code

File: app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt [hunk 13]
```diff
@@ -9,42 +10,89 @@ import androidx.compose.runtime.getValue
...
+                }
+                .addOnFailureListener {
+                  // print the error
+                  println(it.message)
+                  error = it.message.toString()
+                }
           }
-        }) {
+        },
+        modifier = Modifier.semantics { testTag = "signUpB" }) {
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt:88]: Same remark:  `Modifier.semantics { testTag = "signUpB" }` would become `Modifier.testTag("signUpB")`

File: app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt [hunk 14]
```diff
@@ -9,42 +10,89 @@ import androidx.compose.runtime.getValue
...
+
+            functions.useEmulator("10.0.2.2", 5001)
+            // change the region of the function to europe-west6
+
+            functions
+                .getHttpsCallable("oncallFind")
+                .call(data)
+                .addOnSuccessListener { task ->
+                  val result = task.data as? Map<String, String>
+                  println(result)
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt:75]: Not sure if it is a debugging log

File: app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt [hunk 15]
```diff
@@ -9,42 +10,89 @@ import androidx.compose.runtime.getValue
...
+                  if (result?.get("response") == "User is Found") {
+                    loginViewModel.updateUserInfo()
+                  }
+                  navController.navigate(Destinations.ASSOCIATION_PAGE.route)
+                }
+                .addOnFailureListener {
+                  // print the error
+                  println(it.message)
```
Comment (Reza-rgb) [inline app/src/main/java/com/swent/assos/ui/login/SignUpScreen.kt:83]: Not sure if it is a debugging log

File: functions/main.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+)
+from flask import jsonify
+from google.cloud import firestore
+initialize_app()
+
+@https_fn.on_call(region="europe-west6")
+def oncallFind(req: https_fn.Request) -> https_fn.Response:
+    # the request body is a JSON object with a single key "email"
+    email = req.data["email"]
+    print(req)
```
Comment (Reza-rgb) [inline functions/main.py:25]: Delete debugging log

File: functions/main.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+        #       [{'acronym': 'IN-MA2',
+        #       'path': 'EPFL/ETU/IN-S/IN-MA2',
+        #       'name': "Section d'informatique - Master semestre 2",
+        #       'position': 'Etudiant',
+        #       'phoneList': [],
+        #       'officeList': [],
+        #       'rank': 0, 'code': 50038}],
+        # 'profile': 'marin.philippe'}
+        associations = list()
+        print(profile[0])
```
Comment (Reza-rgb) [inline functions/main.py:45]: Delete debugging log

File: functions/main.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        }
+        
+        firestore_client: google.cloud.firestore.Client = firestore.Client()
+        
+        firestore_client.document(f"users/{userID}").set(user)
+        # transform profile insto a json format
+        
+        print(user)
+        print(jsonify(user))
+        print("User added to firestore")
```
Comment (Reza-rgb) [inline functions/main.py:71]: Not sure if it is debugging log

**Review by Reza-rgb — CHANGES_REQUESTED** [review]
Overall this is a great PR ! There are just some minor changes : delete debugging log, commented code and comply with previous code style / implementation

**Review by Reza-rgb — APPROVED** [review]
All the highlighted problems were resolved. Perfect !

**Comment by marinPh** [comment]
corrected the formatting code

**Review by Reza-rgb — APPROVED** [review]
Formatting was applied, great !

**Pull request merged**